rflx/specification/parser.py

```diff
@@ -1,358 +1,981 @@
+    if state.f_identifier.text != state.f_body.f_end_identifier.text:
+        fail(
+            "inconsistent state identifier: "
+            f"{state.f_identifier.text} /= {state.f_body.f_end_identifier.text}",
+            Subsystem.PARSER,
+            Severity.ERROR,
+            location,
+        )
+    transitions = []
+    for t in state.f_body.f_conditional_transitions:
+        transitions.append(create_transition(t, filename))
+    transitions.append(create_transition(state.f_body.f_final_transition, filename))
+    actions = []
+    for a in state.f_body.f_actions:
+        actions.append(create_statement(a, filename))
+    declarations = []
+    for d in state.f_body.f_declarations:
+        declarations.append(create_declaration(d, filename))
+    description = create_description(state.f_description)
+    return model.State(
+        identifier=identifier,
+        transitions=transitions,
+        actions=actions,
+        declarations=declarations,
+        description=description,
+        location=node_location(state, filename),
+    )
+
+
+def __check_session_identifier(session: SessionDecl, filename: Path = None) -> None:
+    if session.f_identifier.text != session.f_end_identifier.text:
+        fail(
+            "inconsistent session identifier: "
+            f"{session.f_identifier.text} /= {session.f_end_identifier.text}",
+            Subsystem.PARSER,
+            Severity.ERROR,
+            node_location(session, filename),
+        )
+
+
+def create_unproven_session(
+    session: SessionDecl,
+    package: ID,
+    types: Sequence[model.Type] = None,
+    filename: Path = None,
+) -> model.UnprovenSession:
+    __check_session_identifier(session, filename)
+    return model.UnprovenSession(
+        model.qualified_type_identifier(create_id(session.f_identifier, filename), package),
+        create_id(session.f_aspects.f_initial, filename),
+        create_id(session.f_aspects.f_final, filename),
+        [create_state(s, filename) for s in session.f_states],
+        [create_declaration(d, filename) for d in session.f_declarations],
+        [create_formal_declaration(p, filename) for p in session.f_parameters],
+        types or [],
+        node_location(session, filename),
+    )
+
+
+def create_session(
+    session: SessionDecl,
+    package: ID,
+    types: Sequence[model.Type] = None,
+    filename: Path = None,
+) -> model.Session:
+    __check_session_identifier(session, filename)
+    return model.Session(
+        model.qualified_type_identifier(create_id(session.f_identifier, filename), package),
+        create_id(session.f_aspects.f_initial, filename),
+        create_id(session.f_aspects.f_final, filename),
+        [create_state(s, filename) for s in session.f_states],
+        [create_declaration(d, filename) for d in session.f_declarations],
+        [create_formal_declaration(p, filename) for p in session.f_parameters],
+        types or [],
+        node_location(session, filename),
+    )
```

Comment:
`create_unproven_session` is only used by test code, which doesn't make much sense. Both functions are nearly identical, so it would be better to remove the code duplication. One option would be to use `create_unproven_session` in `create_session` and add a `proven` method to `UnprovenSession` which returns a `Session` object (similar as it is done for `UnprovenMessage` and `Message`).

Reply:
I deliberately avoided the `UnprovenMessage` solution. What bothers me with it is that you can just create a proven message by calling the `Message` constructor with `skip_prove` set to true. It would be better to avoid constructively that a proven object can be created without actually running the proofs. I agree that having a `proven` method would be good. Here is my take on that (open for discussions): a0627458

Reply:
I see your concern about the `skip_proof` flag of `Message`. Although that flag is only used in tests for performance reasons, having the possibility to skip checks in production code is not ideal, and we should think about another solution. Nevertheless, I do not understand why you see the same problem for `Session` where such flag does not exist and is not required. As far as I can see the `proven` method of `Message` could be applied to `Session` without the `skip_proof` flag. The solution in 22f8310 has multiple dirty hacks which seem unnecessary to me.

Reply:
Change in 6d7f7c72. I'm somewhat uncomfortable with calling the constructor twice just to get the proven variant of the class. No big deal for `Session`, but if the constructor did more elaborate things, this could be an issue (I guess that's the reason for `skip_proof`).